src/order/filter/basic.lean

```diff
@@ -776,27 +776,48 @@ def seq (f : filter (α → β)) (g : filter α) : filter β :=
     ⟨t₀ ∩ u₀, inter_mem_sets ht₀ hu₀, t₁ ∩ u₁, inter_mem_sets ht₁ hu₁,
       assume x ⟨hx₀, hx₁⟩ x ⟨hy₀, hy₁⟩, ⟨ht _ hx₀ _ hy₀, hu _ hx₁ _ hy₁⟩⟩⟩
 
-instance : has_pure filter := ⟨λ(α : Type u) x, principal {x}⟩
+/-- `pure x` is the set of sets that contain `x`. It is equal to `principal {x}` but
+this definition we have `s ∈ pure a` defeq `a ∈ s`. -/
+instance : has_pure filter := ⟨λ(α : Type u) x,
+  { sets := {s | x ∈ s},
+    inter_sets := λ s t, and.intro,
+    sets_of_superset := λ s t hs hst, hst hs,
+    univ_sets := trivial }⟩
 
 instance : has_bind filter := ⟨@filter.bind⟩
 
 instance : has_seq filter := ⟨@filter.seq⟩
 
 instance : functor filter := { map := @filter.map }
 
+lemma pure_sets (a : α) : (pure a : filter α).sets = {s | a ∈ s} := rfl
+
+@[simp] lemma mem_pure_sets {a : α} {s : set α} : s ∈ (pure a : filter α) ↔ a ∈ s := iff.rfl
+
+lemma pure_eq_principal (a : α) : (pure a : filter α) = principal {a} :=
+filter.ext $ λ s, by simp only [mem_pure_sets, mem_principal_sets, singleton_subset_iff]
+
+@[simp] lemma map_pure (f : α → β) (a : α) : map f (pure a) = pure (f a) :=
+filter.ext $ λ s, iff.rfl
+
+@[simp] lemma join_pure (f : filter α) : join (pure f) = f := filter.ext $ λ s, iff.rfl
+
+@[simp] protected lemma bind_pure (m : α → filter β) (a : α) :
```

Comment:
Isn't this lemma called `pure_bind` everywhere else?

Reply:
What about other lemmas? Should they be `map_pure` or `pure_map`, `join_pure` or `pure_join`?

Reply:
The names are like that because `bind` is an infix operator (just like `add`).  The lemma here is `(pure a >>= m) = m a`.  Read left-to-right, the operators are `pure` and `bind`, hence `pure_bind`.